[core/iceberg] Added optional snapshot summary fields to iceberg metadata

0dunay0 opened the pull request:
<!-- Please specify the module before the PR name: [core] ... or [flink] ... -->

### Purpose

Linked issue: close #6502 

This PR fixes Redshift Spectrum querying for Paimon tables with Iceberg compatibility by populating optional snapshot summary fields that are required by certain Iceberg query engines.

When Paimon generates Iceberg metadata, it currently only includes the `operation` field in snapshot summaries. While the Iceberg specification marks most summary fields as "optional," some query engines (notably AWS Redshift Spectrum) require fields like `total-records` to successfully parse and query tables.

This causes Paimon+Iceberg tables to be queryable in AWS Athena but fail in Redshift Spectrum with error: `Required field total-records missing`.

## Changes

### Added `computeSnapshotSummary()` Helper Method

Aggregates statistics from `IcebergManifestFileMeta` objects to compute snapshot-level metrics including:

**Required fields (always present):**
- `total-records` - Total number of live records
- `total-data-files` - Total number of live data files  
- `total-delete-files` - Total number of live delete files
- `total-position-deletes` - Total position delete records
- `total-equality-deletes` - Always "0" (Paimon doesn't use equality deletes)

**Optional fields (when non-zero):**
- `added-data-files`, `added-records`, `added-files-size`
- `deleted-data-files`, `deleted-records`, `deleted-files-size`
- `total-files-size`
- `changed-partition-count`

### Tests

Updated `IcebergMetadataTest.java`

### API and Format

N/A

### Documentation

Reintroduces a feature that was previously available.

```
aws s3 cp s3://some-bucket/paimon/warehouse/somedb.db/some_table/metadata/v190.metadata.json - | jq '.snapshots[0].summary'

{
  "added-data-files": "2",
  "total-equality-deletes": "0",
  "added-records": "83282",
  "deleted-data-files": "0",
  "deleted-records": "0",
  "total-records": "83282",
  "deleted-files-size": "0",
  "changed-partition-count": "1",
  "total-position-deletes": "0",
  "added-files-size": "4683766",
  "total-delete-files": "0",
  "total-files-size": "4683766",
  "total-data-files": "2",
  "operation": "append"
}
```

Redshift Spectrum can now query the table.

0dunay0 committed:
Added optional snapshot summary fields to iceberg metadata

rajagopal-ravikumar reviewed (approved):
looks good to me

plusplusjiajia commented on paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergCommitCallback.java:
```diff
@@ -1230,6 +1328,147 @@ private List<IcebergManifestFileMeta> createDvManifestFileMetas(Snapshot snapsho
+                .sum();
+    }
+
+    private long computeDeleteRowCount(List<IcebergManifestFileMeta> manifestMetas) {
```
computeLiveRowCount is the same with computeDeleteRowCount now, please check them.

0dunay0 replied:
I removed the duplicate and renamed the shared method to `computeLiveFileCount`. I also found that `total-records` was using `paimonSnapshot.totalRecordCount()` which includes level-0 files that Iceberg can't see in PK tables. Switched it to the same incremental pattern used by `total-data-files` and `total-files-size`.

0dunay0 committed:
Used constants for metrics keys

0dunay0 committed:
Address review comments: deduplicate helper methods, fix total-records accuracy

- Removed duplicate computeLiveDeleteFileCount (identical to computeLiveDataFileCount),
  renamed to computeLiveFileCount since both operate on pre-filtered manifest lists.
- Fixed total-records in snapshot summary to track Iceberg-visible records instead of
  using paimon's totalRecordCount which includes non-visible level-0 files in PK tables.
- Adapted to Snapshot.totalRecordCount() changing from Long to long on master.

0dunay0 commented on paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergCommitCallback.java:
```diff
@@ -1230,6 +1328,147 @@ private List<IcebergManifestFileMeta> createDvManifestFileMetas(Snapshot snapsho
+        long totalPositionDeletes = Math.max(0, metrics.totalPositionDeletes);
+        long totalEqualityDeletes = Math.max(0, metrics.totalEqualityDeletes);
+
+        summary.put("added-data-files", Long.toString(addedDataFiles));
```
Added the metrics as constants.

0dunay0 commented:
@plusplusjiajia please have a look on this again

0dunay0 commented:
@JingsongLi  please have a look

0dunay0 commented:
@JingsongLi / @plusplusjiajia / @Zouxxyy / @yuzelin Can you have a look at this PR please?

JingsongLi reviewed (commented):
Good fix for cross-engine compatibility. Redshift Spectrum requiring `total-records` despite Iceberg spec marking it optional is a real-world issue.

**Review:**

1. **Approach**: Computing snapshot summary statistics from `IcebergManifestFileMeta` is correct — these manifests already contain the per-file statistics needed to compute totals.

2. **Fields added**: `total-records`, `total-data-files`, `total-delete-files`, `total-position-deletes`, `total-equality-deletes` (always "0") — these cover Redshift Spectrum's requirements. Adding `added-*` and `deleted-*` fields when non-zero is a nice extra.

3. **291 additions** including test changes — reasonable for the scope.

4. **Correctness concern**: The `total-records` should represent the count of *live* records (not including deleted). Ensure the computation subtracts records in delete files from the total.

5. **Performance**: Computing summary stats adds overhead to each Iceberg commit. For tables with many manifest files, this could be noticeable. Is the computation incremental (building on previous snapshot's summary) or full recomputation?

6. **Test coverage**: `IcebergCompatibilityTest` and `IcebergMetadataTest` updated — good.

LGTM. This unblocks a real user scenario (Paimon + Iceberg compat + Redshift Spectrum).

JingsongLi commented:
+1
